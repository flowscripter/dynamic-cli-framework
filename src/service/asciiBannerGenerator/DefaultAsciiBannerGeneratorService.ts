import type AsciiBannerGeneratorService from "../../api/service/core/AsciiBannerGeneratorService.ts";
import figlet from "figlet";
// FIGlet font is converted to a JSON string and embedded in a simple JSON file: `{ "font": "<figlet font definition>" }`
import standardFont from "./standard.flf.json" with { type: "json" };

/**
 * Default implementation of {@link AsciiBannerGeneratorService} which has a font definition
 * for the FIGlet "standard" font already registered.
 */
export default class DefaultAsciiBannerGeneratorService
  implements AsciiBannerGeneratorService {
  #registeredFontNames: Array<string> = [];

  constructor() {
    this.registerFont("standard", standardFont.font);
  }
  getRegisteredFonts(): ReadonlyArray<string> {
    return this.#registeredFontNames;
  }

  registerFont(fontName: string, fontDefinition: string): void {
    const name = fontName.toLowerCase();
    if (this.getRegisteredFonts().includes(name)) {
      throw new Error(`Font name already registered: ${name}`);
    }
    this.#registeredFontNames.push(name);
    figlet.parseFont(fontName, fontDefinition);
  }

  async generate(message: string, fontName: string): Promise<string> {
    const name = fontName.toLowerCase();
    if (!this.getRegisteredFonts().includes(name)) {
      throw new Error(`Font name is not registered: ${name}`);
    }

    return "\n" + await figlet.text(message, { font: name }) + "\n";
  }
}
