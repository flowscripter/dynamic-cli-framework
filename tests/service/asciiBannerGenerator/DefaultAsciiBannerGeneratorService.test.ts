import { assertEquals, assertRejects, assertThrows } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import DefaultAsciiBannerGeneratorService from "../../../src/service/asciiBannerGenerator/DefaultAsciiBannerGeneratorService.ts";

// FIGlet font is converted to a JSON string and embedded in a simple JSON file: `{ "font": "<figlet font definition>" }`
import smallFont from "./small.flf.json" with { type: "json" };

Deno.test("standard font registered by default", () => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  assertEquals(asciiBannerGeneratorService.getRegisteredFonts(), [
    "standard",
  ]);
});

Deno.test("Cannot register a font if already registered", () => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  assertThrows(() =>
    asciiBannerGeneratorService.registerFont("standard", smallFont.font)
  );
});

Deno.test("Can register new font", () => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  asciiBannerGeneratorService.registerFont("small", smallFont.font);
  assertEquals(asciiBannerGeneratorService.getRegisteredFonts(), [
    "standard",
    "small",
  ]);
});

Deno.test("Cannot generate with unknown font", async () => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  await assertRejects(() =>
    asciiBannerGeneratorService.generate("foo", "small")
  );
});

Deno.test("Can generate with standard font", async (t) => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  asciiBannerGeneratorService.registerFont("small", smallFont.font);

  const generated = await asciiBannerGeneratorService.generate(
    "foo",
    "standard",
  );
  await assertSnapshot(t, generated);
});

Deno.test("Can generate with newly registered font", async (t) => {
  const asciiBannerGeneratorService = new DefaultAsciiBannerGeneratorService();

  asciiBannerGeneratorService.registerFont("small", smallFont.font);

  const generated = await asciiBannerGeneratorService.generate(
    "foo",
    "small",
  );
  await assertSnapshot(t, generated);
});
