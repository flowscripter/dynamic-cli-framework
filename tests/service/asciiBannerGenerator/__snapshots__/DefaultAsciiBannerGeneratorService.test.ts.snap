export const snapshot = {};

snapshot[`Can generate with standard font 1`] = `
"
   __             
  / _| ___   ___  
 | |_ / _ \\\\ / _ \\\\ 
 |  _| (_) | (_) |
 |_|  \\\\___/ \\\\___/ 
                  
"
`;

snapshot[`Can generate with newly registered font 1`] = `
"
   __          
  / _|___  ___ 
 |  _/ _ \\\\/ _ \\\\
 |_| \\\\___/\\\\___/
               
"
`;
